server/migrate/Import.java

```diff
@@ -0,0 +1,420 @@
+                }
+
+                if (player != null) {
+                    if (role == null) {
+                        role = roleCache.computeIfAbsent(roleMessage.getLabel(), l -> currentTransaction.getRole(l));
+                    }
+                    relation.assign(role, player);
+                    roleCount++;
+                } else {
+                    missingPlayers.add(originalPlayerId);
+                }
+            }
+
+            if (!missingPlayers.isEmpty()) {
+                missingRolePlayers.add(new Pair<>(roleMessage.getLabel(), missingPlayers));
+            }
+        }
+
+        if (!missingRolePlayers.isEmpty()) {
+            this.missingRolePlayers.add(new Pair<>(importedId, missingRolePlayers));
+        }
+
+        insertOwnedAttributesThatExist(relation, relationMessage.getAttributeList());
+
+        relationCount++;
+        write();
+    }
+
+    private void insertOwnedAttributesThatExist(Thing thing, List<DataProto.Item.OwnedAttribute> owned) {
+        List<String> missingOwnerships = new ArrayList<>();
+
+        for (DataProto.Item.OwnedAttribute ownedMessage : owned) {
+            String ownedOriginalId = ownedMessage.getId();
+            Attribute<?> ownedAttribute = (Attribute<?>) thingCache.get(ownedOriginalId);
+            if (ownedAttribute == null) {
+                String ownedNewId = idMap.get(ownedOriginalId);
+                if (ownedNewId != null) {
+                    ownedAttribute = currentTransaction.getConcept(ConceptId.of(ownedNewId));
+                    thingCache.put(ownedOriginalId, ownedAttribute);
+                }
+            }
+
+            if (ownedAttribute != null) {
+                thing.has(ownedAttribute);
+                ownershipCount++;
+            } else {
+                missingOwnerships.add(ownedMessage.getId());
+            }
+        }
+
+        if (!missingOwnerships.isEmpty()) {
+            missingAttributeOwnerships.add(new Pair<>(thing.id().toString(), missingOwnerships));
+        }
+    }
+
+    private void insertMissingAttributeOwnershipsAtEnd() {
+        for (Pair<String, List<String>> attributeKeyOwnership : missingAttributeOwnerships) {
+            Attribute<?> owner = currentTransaction.getConcept(ConceptId.of(attributeKeyOwnership.first()));
+
+            for (String attributeIdString : attributeKeyOwnership.second()) {
+                Attribute<?> owned = (Attribute<?>) thingCache.computeIfAbsent(
+                        idMap.get(attributeIdString), // Second is foreign ID, remember to map
+                        id -> currentTransaction.getConcept(ConceptId.of(id)));
+                owner.has(owned);
+                ownershipCount++;
+            }
+
+            write();
+        }
+
+        missingAttributeOwnerships.clear();
+    }
+
+    private void insertMissingRolePlayersAtEnd() {
+        for (Pair<String, List<Pair<String, List<String>>>> missingRolePlayers : missingRolePlayers) {
```

Comment:
Yes it does, I'm not a huge fan of the name duplication so I will change it, but the reason this works is because local variables can shadow class fields (they just can't shadow other local variables).